simSPI/linear_simulator/ctf_utils.py

```diff
@@ -0,0 +1,193 @@
+
+                defocus_u: horizontal defocus in um (comes from ctf_params);
+                defocus_v: vertical defocus in um (comes from ctf_params);
+                defocus_angle: astigmatism angle in radians (comes from ctf_params).
+
+            For a given dataset alpha_f, ac, wavelength, and cs are constants,
+            whereas, other variables can vary per projection.
+
+        Discretization
+        --------------
+            The code below simplifies and discretizes the above continuous-domain
+            expression by sampling CTF value at config.ctf_size//2 number of
+            equidistant frequency samples between 0 to Nyquist frequency.
+
+            Specifically, in a grid of pixel location going from (0,0) to (N-1, N-1)
+            for any pixel location (m,n):
+
+            |f|= ((m-c)^2+(n-c)^2)**0.5 *frequency_step,and
+            alpha_f=arctan(n-c,m-c)
+            where frequency_step= Nyquist_frequency/(N/2)=1/(pixel_size*N)
+
+        References
+        ----------
+            [1]: Frank, Joachim. Three-dimensional electron microscopy of
+                macromolecular assemblies: visualization of biological molecules
+                in their native state. Oxford university press, 2006.
+            [2]: Scheres, Sjors HW. "RELION: implementation of a Bayesian approach to
+                cryo-EM structure determination." Journal of structural biology,
+                180, no. 3 (2012): 519-530.
+
+        Remarks
+        -------
+            In the code below, the envelope function E(f), when b_factor is unknown
+            (assumed to be 0 in that case) uses value_nyquist given by the user. This
+            corresponds to the value of envelope function at nyquist frequency and
+            hence, is directly related to CTF b_factor. Because of this simple
+            definition of value_nyquist it is a more intuitive and easier input
+            for the user to provide than b_factor.
+
+        Parameters
+        ----------
+        ctf_params: dict of type str to torch.Tensor
+            contains defocus parameters with keys
+                defocus_u: str to torch.Tensor
+                    horizontal defocus in um with shape: (batch_size,1,1,1)
+                defocus_v: str to torch.Tensor
+                    vertical defocus in um with shape: (batch_size,1,1,1)
+                defocus_angle: str to torch.Tensor
+                    astigmatism angle in radians with shape: (batch_size,1,1,1)
+
+        Returns
+        -------
+        hFourier: torch.Tensor
+            ctf of shape (batch_size,1,ctf_size,ctf_size)
+        """
+        defocus_u = ctf_params["defocus_u"]
+        defocus_v = ctf_params["defocus_v"]
+        angle_astigmatism = ctf_params["defocus_angle"]
+        elliptical = (
+            defocus_v
+            + (defocus_u - defocus_v)
+            * torch.cos(self.angleFrequency - angle_astigmatism) ** 2
+        )
+        defocusContribution = (
```

Comment:
no CamelCase in python, see pep8 https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

Reply:
Sorry missed them. Corrected. Thanks!